staging/src/k8s.io/client-go/util/flowcontrol/throttle.go

```diff
@@ -39,20 +38,42 @@ type RateLimiter interface {
 	Wait(ctx context.Context) error
 }
 
-type tokenBucketRateLimiter struct {
+type RateLimiter interface {
+	PassiveRateLimiter
+	// Accept returns once a token becomes available.
+	Accept()
+}
+
+type clockIndependentRateLimiter struct {
 	limiter *rate.Limiter
-	clock   Clock
 	qps     float32
 }
 
+type tokenBucketRateLimiter struct {
+	clockIndependentRateLimiter
```

Comment:
@MikeSpreitzer I'm not sure about this, mainly because we would still need to implement the same number of distinct methods (I don't think `tokenBucketRateLimiter` can use `tokenBucketPassiveRateLimiter`'s `TryAccept()` because of the type of clocks used in the two).

Perhaps I misunderstood, please let me know if I'm missing anything

Reply:
I am wondering along the lines of the following.

```
type tokenBucketPassiveRateLimiter struct {
	limiter *rate.Limiter
	clock   clock.PassiveClock
	qps     float32
}


type tokenBucketRateLimiter struct {
	tokenBucketPassiveRateLimiter
	clock Clock
}

 func newTokenBucketRateLimiterWithClock(limiter *rate.Limiter, c Clock, qps float32) RateLimiter {
	return &tokenBucketRateLimiter{
		tokenBucketPassiveRateLimiter: newTokenBucketRateLimiterWithPassiveClock(limiter, c, qps),
		clock: c}
}

 func newTokenBucketRateLimiterWithPassiveClock(limiter *rate.Limiter, c clock.PassiveClock, qps float32) *tokenBucketPassiveRateLimiter {
...
}
```

Reply:
Ahh I see, thanks @MikeSpreitzer!
I've made the changes, `Clock` now extends `clock.PassiveClock` as well in order to simplify things further :)

Reply:
Yes, I was expecting you to write

```
type Clock interface {
	clock.PassiveClock
	Sleep(time.Duration)
}
```
and use `clock.RealClock` instead of defining another RealClock here.  Does that not solve the problem?

Remember also that Wait is not passive.

Reply:
@MikeSpreitzer that's the change I made initially. But the only issue with this is that other parts of the codebase implement `flowcontrol.Clock`, and that would lack the `Since()` method. 
https://github.com/kubernetes/kubernetes/blob/b12379ef1c8defd0bcda12103a46b05ab094a641/plugin/pkg/admission/eventratelimit/limitenforcer.go#L45

One thing we could do is refactor things here as well to use `clock.RealClock` instead of a newly defined `realClock` or implement a `Since` method for this defined `realClock`

Reply:
I am not sure I understand.  Both definitions of `PassiveClock` have the `Since` method:
- https://github.com/kubernetes/apimachinery/blob/74be3b88bedbf44de881c8c33c5b1c3c536ad1f4/pkg/util/clock/clock.go#L29
- https://github.com/kubernetes/utils/blob/d34e5cb4466e533a699c4cdbd76f3a8220b21d54/clock/clock.go#L26

Remember that Wait is not passive; it can not be in the passive rate limiter types/interfaces.

Reply:
@MikeSpreitzer 

- If `clock.PassiveClock` is extended through `flowcontrol.Clock`, then the resultant would be
```go
// package flowcontrol
type Clock interface {
    clock.PassiveClock
    Sleep(time.Duration)
}
```
meaning that any `clock` that were to implement `flowcontrol.Clock` would have to implement the `Now()`, `Since()` and `Sleep()` methods.

The issue that I tried to convey was that in places such as:
https://github.com/kubernetes/kubernetes/blob/b12379ef1c8defd0bcda12103a46b05ab094a641/plugin/pkg/admission/eventratelimit/limitenforcer.go#L45

it is called as follows:
https://github.com/kubernetes/kubernetes/blob/b12379ef1c8defd0bcda12103a46b05ab094a641/plugin/pkg/admission/eventratelimit/admission.go#L50

and `realClock` here implements `Now` and `Sleep`, but not `Since`, which is why it won't be an implementation of `flowcontrol.Clock`.
https://github.com/kubernetes/kubernetes/blob/b12379ef1c8defd0bcda12103a46b05ab094a641/plugin/pkg/admission/eventratelimit/clock.go#L24

---

> Remember that Wait is not passive; it can not be in the passive rate limiter types/interfaces.

👍🏻 
So the rate limiting interfaces would now look like this?
```go
type PassiveRateLimiter interface {
	// TryAccept returns true if a token is taken immediately. Otherwise,
	// it returns false.
	TryAccept() bool
	// Stop stops the rate limiter, subsequent calls to CanAccept will return false
	Stop()
	// QPS returns QPS of this rate limiter
	QPS() float32
}

type RateLimiter interface {
	PassiveRateLimiter
	// Accept returns once a token becomes available.
	Accept()
	// Wait returns nil if a token is taken before the Context is done.
	Wait(ctx context.Context) error
}
```

Reply:
(A) Thanks for the explanation, now I understand what you were trying to say about the `Since` method.  I see that in the latest revision, you have added the `Since` method to the local `realClock`.  FYI, in case you have not noticed, there are already `realClock` types in the two `clock` packages that implement all the methods you need.

(B) Regarding where to put the `Wait` method: yes, what you outlined looks good.